source/common/config/unified_mux/delta_subscription_state.cc

```diff
@@ -0,0 +1,233 @@
+    // Also, since this might be a new server, we must explicitly state *all* of our subscription
+    // interest.
+    for (auto const& [resource_name, resource_state] : resource_state_) {
+      // Populate initial_resource_versions with the resource versions we currently have.
+      // Resources we are interested in, but are still waiting to get any version of from the
+      // server, do not belong in initial_resource_versions. (But do belong in new subscriptions!)
+      if (!resource_state.isWaitingForServer()) {
+        (*request->mutable_initial_resource_versions())[resource_name] = resource_state.version();
+      }
+      // As mentioned above, fill resource_names_subscribe with everything, including names we
+      // have yet to receive any resource for unless this is a wildcard subscription, for which
+      // the first request on a stream must be without any resource names.
+      if (!wildcard_) {
+        names_added_.insert(resource_name);
+      }
+    }
+    // Wildcard subscription initial requests must have no resource_names_subscribe.
+    if (wildcard_) {
+      names_added_.clear();
+    }
+    names_removed_.clear();
+  }
+
+  std::copy(names_added_.begin(), names_added_.end(),
+            Protobuf::RepeatedFieldBackInserter(request->mutable_resource_names_subscribe()));
+  std::copy(names_removed_.begin(), names_removed_.end(),
+            Protobuf::RepeatedFieldBackInserter(request->mutable_resource_names_unsubscribe()));
+  names_added_.clear();
+  names_removed_.clear();
+
+  return request;
+}
+
+envoy::service::discovery::v3::DeltaDiscoveryRequest*
+DeltaSubscriptionState::getNextRequestAckless() {
+  return getNextRequestInternal();
+}
+
+envoy::service::discovery::v3::DeltaDiscoveryRequest*
+DeltaSubscriptionState::getNextRequestWithAck(const UpdateAck& ack) {
+  auto* request = getNextRequestInternal();
+  request->set_response_nonce(ack.nonce_);
+  ENVOY_LOG(debug, "ACK for {} will have nonce {}", type_url(), ack.nonce_);
+  if (ack.error_detail_.code() != Grpc::Status::WellKnownGrpcStatus::Ok) {
+    // Don't needlessly make the field present-but-empty if status is ok.
+    request->mutable_error_detail()->CopyFrom(ack.error_detail_);
+  }
+  return request;
+}
+
+void DeltaSubscriptionState::addResourceState(
+    const envoy::service::discovery::v3::Resource& resource) {
+  if (resource.has_ttl()) {
+    ttl_.add(std::chrono::milliseconds(DurationUtil::durationToMilliseconds(resource.ttl())),
+             resource.name());
+  } else {
+    ttl_.clear(resource.name());
+  }
+
+  resource_state_[resource.name()] = ResourceState(resource.version());
+}
+
+void DeltaSubscriptionState::ttlExpiryCallback(const std::vector<std::string>& expired) {
+  Protobuf::RepeatedPtrField<std::string> removed_resources;
+  for (const auto& resource : expired) {
+    resource_state_[resource] = ResourceState::waitingForServer();
+    removed_resources.Add(std::string(resource));
```

Comment:
Why is the call to the string c'tor required here?

Reply:
This is a port from here: https://github.com/envoyproxy/envoy/blob/main/source/common/config/delta_subscription_state.cc#L22. My understaning is that `Protobuf::RepeatedPtrField` takes ownership of strings/massages added to it, hence the signature `(RepeatedPtrField<T>::Add(T&& value))` and the need for a new instance of the string.